internal/controller/base_controller.go

```diff
@@ -0,0 +1,137 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/ngrok/ngrok-api-go/v5"
+	"github.com/ngrok/ngrok-operator/internal/util"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// BaseControllerOp is an enum for the different operations that can be performed by a BaseController
+type BaseControllerOp int
+
+const (
+	// createOp is the operation for creating a resource
+	CreateOp BaseControllerOp = iota
+
+	// updateOp is the operation for updating a resource (upsert)
+	UpdateOp
+
+	// deleteOp is the operation for deleting a resource (and finalizers)
+	DeleteOp
+)
+
+// BaseController is our standard pattern for writing controllers
+//
+// Note: Non-provided methods are not called during reconcile
+type BaseController[T client.Object] struct {
```

Comment:
tbqh I think this would be better suited as an interface and embedded struct types for the consumers, but that's a much larger refactor so I'll come back to that when I'm more familiar with the codebase.